collector/uptrack.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2016 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"flag"
+	"io/ioutil"
+	"path"
+	"syscall"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"gopkg.in/yaml.v2"
+)
+
+var (
+	uptrackUpgradePlanPath = flag.String(
+		"collector.uptrack.upgrade_plan_path",
+		uptrackDefaultUpgradePlanPath(),
+		"Path where Uptrack stores its upgrade_plan YAML file.")
+)
+
+type uptrackCollector struct {
+	timeDesc      *prometheus.Desc
+	planCountDesc *prometheus.Desc
+}
+
+func init() {
+	Factories["uptrack"] = NewUptrackCollector
+}
+
+// The location of the upgrade_plan file is based on the currently running kernel.
+func uptrackDefaultUpgradePlanPath() string {
+	var uname syscall.Utsname
+	syscall.Uname(&uname)
+	return path.Join(
+		"/var/cache/uptrack",
```

Comment:
I'd say that even for per-distro things, running a separate exporter for every tiny component can easily become messy. This is why I think that having a collector for this in the node exporter makes a lot of sense. From a performance/efficiency point of view there is little to no overhead in having collectors present that are disabled by default, right?

A colleague of mine is currently developing a Puppet state collector for the node exporter for the same reasons, for which he is planning on filing a pull request later this week.